### ISSUE TITLE
[RFC]: Optional NCCL m2n Reshard for Non-Colocated Weight Sync

### ISSUE THREAD

**Issue by princepride** [vllm-project/vime#52]
## Propose

This RFC proposes adding NVIDIA's experimental NCCL Xfer Reshard primitive as an optional non-colocated rollout weight update path.

Today, vime updates non-colocated rollout engines with a gather-broadcast flow: Megatron shards are gathered on selected trainer ranks, converted to vLLM/HF tensors, then broadcast to rollout ranks. The proposed design keeps this existing path as the default and adds an opt-in path that directly reshards each logical tensor from the trainer layout to the rollout layout with `ncclXferReshardWithWindow`.

Colocated rollout engines should keep using the existing CUDA IPC path. This RFC only targets non-colocated rollout weight sync.

References:

- NVIDIA NCCL m2n source: <https://github.com/NVIDIA/nccl/tree/v2.30u1/contrib/nccl_m2n>
- Public API: `nccl_xfer/src/nccl_xfer.h`
- Model benchmark: `nccl_xfer/benchmarks/reshard_model_bench.cu`

## Motivation

The current gather-broadcast path is simple and mature, but it can create avoidable traffic and hot spots:

- Trainer TP/EP shards are first gathered into full tensors on source ranks.
- Source ranks then broadcast payloads to all rollout ranks.
- The transfer does not express the real operation, which is a layout change from trainer sharding to rollout sharding.
- Replicated rollout layouts can create redundant network traffic.

NCCL Xfer Reshard is designed for this shape: two disjoint GPU groups hold the same logical tensor in different distributed layouts, and one collective moves data from the source layout to the destination layout. Because this is a new and experimental NCCL contrib feature, it should be introduced as an opt-in strategy before becoming a default.

<img width="682" height="318" alt="Image" src="https://github.com/user-attachments/assets/76ea57d6-c8b2-44c6-9270-123163154e54" />

<img width="1105" height="484" alt="Image" src="https://github.com/user-attachments/assets/4a52f6c4-5be8-4aee-a4c0-b5e7e168c022" />

## Current Path

```text
Megatron TP/EP shards
  -> trainer-side all-gather
  -> Megatron-to-HF conversion
  -> trainer source rank
  -> NCCL broadcast via vLLM weight transfer
  -> rollout ranks
```

Relevant code:

- `UpdateWeightFromDistributed` in `vime/backends/megatron_utils/update_weight/update_weight_from_distributed.py`
- `all_gather_param` in `vime/backends/megatron_utils/update_weight/common.py`
- vLLM payload transfer through `NCCLWeightTransferEngine.trainer_send_weights`

This path should remain available as the stable default and as the fallback for unsupported NCCL Xfer cases.

## Proposed Optional Path

```text
Megatron trainer distributed tensor layout
  -> NCCL Xfer Reshard
  -> rollout distributed tensor layout
  -> local vLLM weight load
```

At a high level:

1. Build one NCCL communicator containing trainer source ranks and rollout destination ranks.
2. Describe the trainer layout with `ncclXferReshardMesh_t` and `ncclXferDistTensor_t`.
3. Describe the rollout layout with the same NCCL Xfer descriptors.
4. Register source and destination buffers with NCCL user windows.
5. Call `ncclXferReshardWithWindow`.
6. Load the received rollout-local shard into vLLM.

The updater selection should become explicit:

```python
if args.colocate:
    update_weight_cls = UpdateWeightFromTensor
elif args.non_colocate_weight_sync_backend == "nccl-xfer":
    update_weight_cls = UpdateWeightFromNcclXfer
else:
    update_weight_cls = UpdateWeightFromDistributed
```

Suggested user-facing option:

```text
--non-colocate-weight-sync-backend {broadcast,nccl-xfer}
```

The default should be `broadcast` initially. `nccl-xfer` should be opt-in until its correctness, compatibility, and performance are validated across dense, PP, MoE, and quantized workloads.

## NCCL Xfer API Shape

```c
typedef struct {
  int dims[2];
  int startRank;
  int placement[2];
} ncclXferReshardMesh_t;

#define NCCLXFER_RESHARD_REPLICATE (-1)
#define NCCLXFER_RESHARD_SHARD(td) (td)

typedef struct {
  void* dataPtr;
  size_t localShape[NCCLXFER_RESHARD_MAX_TENSOR_DIMS];
  int ndims;
  ncclDataType_t dtype;
  const ncclXferReshardMesh_t* mesh;
} ncclXferDistTensor_t;

ncclResult_t ncclXferReshardWithWindow(
    ncclComm_t comm,
    ncclWindow_t window,
    const ncclXferDistTensor_t* src,
    const ncclXferDistTensor_t* dst,
    cudaStream_t stream);
```

Important properties:

- Source and destination ranks live in the same NCCL communicator.
- Ranks that do not participate on one side pass `dataPtr = NULL` for that side.
- Buffers must be registered as NCCL user windows.
- The library supports `RING` and `DIRECT`; for this use case we should start with `RING` / `AUTO` and node-aware load balancing.

## Layout Mapping

The first implementation should support the common model weight layouts:

- Column-parallel tensors: shard tensor dim 0.
- Row-parallel tensors: shard tensor dim 1.
- Embeddings and LM head: usually shard tensor dim 0.
- LayerNorm, bias, router weights: replicated.
- MoE expert weights: group experts into a 3D logical tensor and shard the expert dimension.
- Pipeline parallelism: create communicators per `(trainer_pp_stage, rollout_pp_stage)` pair instead of encoding PP into the NCCL Xfer mesh.

This should reuse existing Megatron-to-HF naming and layout metadata wherever possible.

## Implementation Plan

Add an optional updater:

```text
vime/backends/megatron_utils/update_weight/update_weight_from_nccl_xfer.py
vime/backends/megatron_utils/update_weight/nccl_xfer_layout.py
vime/backends/megatron_utils/update_weight/nccl_xfer_bindings/
vime/backends/vllm_utils/nccl_xfer_worker.py
```

Initial rollout:

1. Validate `nccl_xfer/` tests and benchmarks independently.
2. Add bindings and window management.
3. Implement dense tensor reshard behind `--non-colocate-weight-sync-backend nccl-xfer`.
4. Add PP-stage communicator mapping.
5. Add MoE grouped expert tensors.
6. Fall back to `broadcast` for unsupported tensors or runtime capability failures.
7. Revisit the default only after stability and performance are proven.

## Fallback Policy

The NCCL Xfer path should fall back to the existing broadcast path when:

- Tensor rank is greater than 3.
- The dtype is unsupported by NCCL Xfer.
- The layout cannot be represented by the current two-axis mesh.
- Quantized or packed formats require special handling.
- vLLM weight storage cannot safely participate in NCCL user-window registration.
- NCCL Xfer initialization or communicator setup fails.

Fallbacks must be explicit in logs and metrics so users can tell whether they are actually using NCCL Xfer.

## Risks

NCCL Xfer is experimental in the referenced NCCL v2.30u1 contrib tree. Known constraints include:

- Tensor rank is limited to 3.
- Meshes are two-axis.
- User-window memory and offset rules must be respected.
- Some mesh-size caps are compile-time constants.
- Failure recovery is more complex because trainer and rollout ranks join one collective.
- vLLM may require staging buffers before direct-to-parameter window registration is safe.

## Testing

Correctness:

- Compare rollout logits after NCCL Xfer update against the legacy gather-broadcast path.
- Validate TP mismatch, PP mapping, replicated tensors, and MoE expert tensors.
- Add checksum/debug validation for received rollout buffers.
- Verify fallback behavior and fallback logging.

Performance:

- Compare total update latency against legacy gather-broadcast.
- Measure max-rank latency, bandwidth, memory overhead, and staging-copy cost.
- Compare `RING` vs `DIRECT` and `NODE_AWARE` vs `UNIFORM`.

## Open Questions

- Should rollout receive directly into vLLM parameter storage or use staging buffers first?
- Should Megatron-to-HF conversion happen before reshard, after reshard, or be split by tensor type?
- What minimum NCCL/CUDA versions should this optional path require?
- How should a failed collective recover and fall back without leaving rollout engines in weight-update mode?
- After validation, should `nccl-xfer` become the default for selected model/topology classes?
